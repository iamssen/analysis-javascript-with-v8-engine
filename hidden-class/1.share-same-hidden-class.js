function Class(x) {
  this.x = x;
}

const a = new Class(1);
const b = new Class(1);

console.log(`a and b are shared the same hidden class.`);

console.log('>>> a');
%DebugPrint(a);
console.log('<<<');

console.log('>>> b');
%DebugPrint(b);
console.log('<<<');
