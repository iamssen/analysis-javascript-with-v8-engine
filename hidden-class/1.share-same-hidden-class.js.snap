"back pointer"가 같다. a와 b는 같은 Hidden Class를 공유한다.
>>> a
DebugPrint: 0x531080c8309: [JS_OBJECT_TYPE]
 - map: 0x05310828702d <Map(HOLEY_ELEMENTS)> [FastProperties]
 - prototype: 0x0531080c8285 <Object map = 0x53108287005>
 - elements: 0x0531080426e5 <FixedArray[0]> [HOLEY_ELEMENTS]
 - properties: 0x0531080426e5 <FixedArray[0]> {
    0x5310825201d: [String] in OldSpace: #x: 1 (const data field 0)
 }
0x5310828702d: [Map]
 - type: JS_OBJECT_TYPE
 - instance size: 48
 - inobject properties: 9
 - elements kind: HOLEY_ELEMENTS
 - unused property fields: 8
 - enum length: invalid
 - stable_map
 - back pointer: 0x053108286f65 <Map(HOLEY_ELEMENTS)>
 - prototype_validity cell: 0x053108252419 <Cell value= 0>
 - instance descriptors (own) #1: 0x0531080c8361 <DescriptorArray[1]>
 - prototype: 0x0531080c8285 <Object map = 0x53108287005>
 - constructor: 0x053108252311 <JSFunction Class (sfi = 0x53108252165)>
 - dependent code: 0x0531080421e9 <Other heap object (WEAK_FIXED_ARRAY_TYPE)>
 - construction counter: 6

<<<
>>> b
DebugPrint: 0x531080c8399: [JS_OBJECT_TYPE]
 - map: 0x05310828702d <Map(HOLEY_ELEMENTS)> [FastProperties]
 - prototype: 0x0531080c8285 <Object map = 0x53108287005>
 - elements: 0x0531080426e5 <FixedArray[0]> [HOLEY_ELEMENTS]
 - properties: 0x0531080426e5 <FixedArray[0]> {
    0x5310825201d: [String] in OldSpace: #x: 1 (const data field 0)
 }
0x5310828702d: [Map]
 - type: JS_OBJECT_TYPE
 - instance size: 48
 - inobject properties: 9
 - elements kind: HOLEY_ELEMENTS
 - unused property fields: 8
 - enum length: invalid
 - stable_map
 - back pointer: 0x053108286f65 <Map(HOLEY_ELEMENTS)>
 - prototype_validity cell: 0x053108252419 <Cell value= 0>
 - instance descriptors (own) #1: 0x0531080c8361 <DescriptorArray[1]>
 - prototype: 0x0531080c8285 <Object map = 0x53108287005>
 - constructor: 0x053108252311 <JSFunction Class (sfi = 0x53108252165)>
 - dependent code: 0x0531080421e9 <Other heap object (WEAK_FIXED_ARRAY_TYPE)>
 - construction counter: 6

<<<
